Addition of F5 BigIP iControl REST Execution Module

Anthonyhawkins opened the pull request:
# Summary

Add an execution module which can interact with an F5 load balancer via iControl REST.  This has been tested against a stand-alone BIG-IP 11.6.0 Build 0.0.401 Final.

Commit:
Initial Commit

Initial code commit and function:

list_node
create_node
modify_node
delete_node
list_pool
create_pool
modify_pool
delete_pool
replace_pool_members
add_pool_member
modify_pool_member
delete_pool_member
list_virtual
create_virtual
modify_virtual
delete_virtual
list_monitor
create_monitor
modify_monitor
delete_monitor
list_profile
create_profile
modify_profile
delete_profile

Commit:
Initial Commit

Initial code commit and function:

list_node
create_node
modify_node
delete_node
list_pool
create_pool
modify_pool
delete_pool
replace_pool_members
add_pool_member
modify_pool_member
delete_pool_member
list_virtual
create_virtual
modify_virtual
delete_virtual
list_monitor
create_monitor
modify_monitor
delete_monitor
list_profile
create_profile
modify_profile
delete_profile

Commit:
Initial Commit

Initial code commit and function:

list_node
create_node
modify_node
delete_node
list_pool
create_pool
modify_pool
delete_pool
replace_pool_members
add_pool_member
modify_pool_member
delete_pool_member
list_virtual
create_virtual
modify_virtual
delete_virtual
list_monitor
create_monitor
modify_monitor
delete_monitor
list_profile
create_profile
modify_profile
delete_profile

Commit:
Added virtual function.

Commit:
Added virtual function.

Commit:
Added virtual function.

Commit:
Added virtual function.

Commit:
Added list and dictionary logic for option builders

Added logic to not try and build lists and dictionaries for parameters
if arg passed is already a list / dictionary

Commit:
Added list and dictionary logic for option builders

Added logic to not try and build lists and dictionaries for parameters
if arg passed is already a list / dictionary

Commit:
Added list and dictionary logic for option builders

Added logic to not try and build lists and dictionaries for parameters
if arg passed is already a list / dictionary

Commit:
Added list and dictionary logic for option builders

Added logic to not try and build lists and dictionaries for parameters
if arg passed is already a list / dictionary

Anthonyhawkins committed:
Delete h3ogajn2.544.txt

Anthonyhawkins committed:
Delete h3ogajn2.544.txt

Anthonyhawkins committed:
Delete h3ogajn2.544.txt

Anthonyhawkins committed:
Delete h3ogajn2.544.txt

Anthonyhawkins committed:
Delete sjpzsiva.bnh.txt

Anthonyhawkins committed:
Delete sjpzsiva.bnh.txt

Anthonyhawkins committed:
Delete sjpzsiva.bnh.txt

Anthonyhawkins committed:
Delete sjpzsiva.bnh.txt

Anthonyhawkins committed:
Delete v3qymexn.q1w.txt

Anthonyhawkins committed:
Delete v3qymexn.q1w.txt

Anthonyhawkins committed:
Delete v3qymexn.q1w.txt

Commit:
Merge branch 'develop' of https://github.com/Anthonyhawkins/salt into add-f5-bigip-execution-module

Commit:
bug fixes and profiles rework

fixed numerous bugs and typos in addition to reworking the way profiles
are handled.  Added support for shorthand list, dictionary, lists of
dictionaries and json support for profile creation and modification.

Commit:
bug fixes and profiles rework

fixed numerous bugs and typos in addition to reworking the way profiles
are handled.  Added support for shorthand list, dictionary, lists of
dictionaries and json support for profile creation and modification.

jfindlay commented:
@Anthonyhawkins, awesome, thanks.  There are a few test failures:
https://jenkins.saltstack.com/job/salt-pr-lint-n/4131/violations/file/salt/modules/bigip.py/
https://jenkins.saltstack.com/job/salt-pr-rs-cent7-n/3700/testReport/junit/integration.modules.sysmod/SysModuleTest/test_valid_docs/

thatch45 commented:
@techhat this only uses requests to make the session, we can do that with salt.utils.http and get rid of the requests dep right?

thatch45 commented:
@Anthonyhawkins This is a great addition! Thanks!

Commit:
Merge branch 'develop' of https://github.com/Anthonyhawkins/salt into add-f5-bigip-execution-module

Commit:
Addressed formatting and other failures from tests

addressed issues found through automated testing. modify_profile now has
the same functionality regarding vlans as create_profile.  Added example
for modify_node.  Also changed variables named type to monitor_type and
profile_type respectively.

Anthonyhawkins commented:
@thatch45 @jfindlay 

Thanks for the feedback, the end goal is to create a state module to complement the execution module.  Before starting the state module I wanted to get the execution portion built first and take away any best practices etc. and apply it to the state portion.

Commit:
Indent and formatting issues

thatch45 commented on salt/modules/bigip.py:
```diff
@@ -0,0 +1,1723 @@
+'''
+
+# Import python libs
+import requests
```
Sorry I missed this one, since requests is an optional salt dep (we had to remove it in 2015.2 because of issues introduced with it in 2014.7) we should gate the import and make a virtual function:
https://github.com/saltstack/salt/blob/develop/salt/modules/etcd_mod.py#L31
https://github.com/saltstack/salt/blob/develop/salt/modules/etcd_mod.py#L52

Anthonyhawkins replied:
Ah, didn't realize requests dependency for salt was removed.  Good to know.  It would probably make more sense to utilize salt.utils.http then.

thatch45 replied:
I dont' think it has sessions yet, we need to get that in, I think it only has query still.

I am going to merge this in, we can fix the imports error and deprecate requests out of here later, I don't want this code to go stale just because of a little thing like this!
